### PR TITLE
Update java syllabus with new topics

### PR DESCRIPTION
**Pull request by ruddy-m**
Related issues: [please specify]

## Description:

What are you up to? Fill us in :)

## I solemnly swear that:

- [x] I ran the hugo server and looked at my changed in the browser with my own eyes
- [x] I ran the linter and there were no errors
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code

### CODE REVIEW

File: content/projects/java-specific/functional-interface/_index.md [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+  - topics/java-specific/functional-interface
+ready: true
+submission_type: repo
+tags:
+- functional-interface
+- lambda-expressions
+title: Java Lamba Expressions and Functional Interfaces
+---
+
+### What is the number
```
Comment (sheenarbw) [inline content/projects/java-specific/functional-interface/_index.md]: The html that comes out of this will be:

```
<h1>Java Lamba Expressions and Functional Interfaces</h1>
<h3>What is the number</h3>
```

Can you try to keep the headings in order? h1, then h2 etc? You are probably optimizing for looks right now but looks will change, it will be nice to not have to go back and fix everything.

File: content/projects/java-specific/functional-interface/_index.md [hunk 2]
```diff
@@ -0,0 +1,50 @@
+---
+content_type: project
+flavours:
+- java
+prerequisites:
+  hard:
+  - topic/oop
```
Comment (sheenarbw) [inline content/projects/java-specific/functional-interface/_index.md]: please make sure that these paths are correct. They need to be exactly right.

There isn't a directory called `topic`. And even if I look inside the `topics` directory then there isn't anything named `oop`. Maybe you mean `projects/oop` but that project has type nosubmit. 

So maybe you mean each of the projects inside the oop directory... Can you list out exactly what is required?

File: content/projects/java-specific/functional-interface/_index.md [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+- functional-interface
+- lambda-expressions
+title: Java Lamba Expressions and Functional Interfaces
+---
+
+### What is the number
+
+Write the following methods that return a lambda expression performing a specified action:
+
+1. A method called **isOdd**: The lambda expression must return true if a number is odd or false if it is even.
```
Comment (sheenarbw) [inline content/projects/java-specific/functional-interface/_index.md]: Code formatting uses backticks. Eg

```
A method called `isOdd`
```

File: content/projects/java-specific/functional-interface/_index.md [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+
+```
+ Number is: Odd, NotPalindrome and Prime
+```
+
+***ALL FUNCTIONS MUST BE TESTED***
+
+
+#### Not to reviewer
+- All three function return lambada expressions
```
Comment (sheenarbw) [inline content/projects/java-specific/functional-interface/_index.md]: I might be wrong, but I think we need a blank line above the unordered list, otherwise, it gets printed as a paragraph.

File: content/projects/java-specific/threads/_index.md [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+  soft:
+  - projects/java-specific/functional-interface
+ready: true
+submission_type: repo
+title: Java Threads
+---
+
+### The Chicken and the Farmer
+
+Have you read the book Animal Farm well its not so important for this project but if you havent read it you should; anyways, in this project you are going to create a Chicken and a Farmer each of these will be there own functions and these are the rules
```
Comment (sheenarbw) [inline content/projects/java-specific/threads/_index.md]: This is missing a bit of punctuation. 

Here's a revised version. I also tried to interpret what you meant, I might be wrong. But if I was wrong then the learners are definitely going to get it wrong.

```
Have you read the book Animal Farm? Well, it's not so important for this project, but if you haven't read it you should.  Anyways, in this project, you are going to create two classes,  `Chicken` and  `Farmer`.  They should function as follows:
```

File: content/projects/java-specific/threads/_index.md [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+2. The Chicken should put eggs in a Basket ***(static array/list of size 1)*** and the Farmer should take eggs from the basket
+
+3. The Farmer cannot take from an empty basket
+
+4. The Chicken cannot add to a full basket
+
+5. The basket is full if it has 1 egg
+
+
+That is it, pretty simple. have fun
```
Comment (sheenarbw) [inline content/projects/java-specific/threads/_index.md]: If you call something simple and people struggle then they sometimes get quite discouraged.

Maybe say something like:

```
We kept the instructions open-ended on purpose. This is meant to have a bit of room for interpretation, there is no one perfect solution to this. We are very interested to see how you structure your solution.

Oh yeah, and remember to test your work. Always.
```

File: content/topics/java-specific/concurrency/_index.md [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+content_type: topic
+ready: true
+tags:
+- threads
+- concurrency
+title: Java Multithread, Concurrency and Parallelism
+---
+
+
+I am sure by now you have heard about java being multi-threaded, what does that mean exactly
```
Comment (sheenarbw) [inline content/topics/java-specific/concurrency/_index.md]:
```
I am sure by now you have heard about Java being multi-threaded, what does that mean exactly?
```
It needs a question mark at the end

File: content/topics/java-specific/concurrency/_index.md [hunk 8]
```diff
@@ -0,0 +1,104 @@
...
+
+Unlike for example JavaScript which is a single threaded language Java on the other has multiple threads to help you run your program faster.
```
Comment (sheenarbw) [inline content/topics/java-specific/concurrency/_index.md]: Hmm, this one also needs revision...

There's a trick I use when writing that might help you.  When I write, I imagine what the words will sound like if I read them aloud.  The sentence above has no punctuation, it would be spoken in a single rushed breath.

File: content/topics/java-specific/concurrency/_index.md [hunk 9]
```diff
@@ -0,0 +1,104 @@
...
+
+#### What is Multithreading
+
+Multithreading means that you have multiple threads of execution inside the same application. A thread is like a separate CPU executing your application. Thus, a multithreaded application is like an application that has multiple CPUs executing different parts of the code at the same time. 
+
+***A thread goes through a life-cycle which can be seen in the diagram below***
+
+![thread-life-cycle](Thread_Life_Cycle.jpg)
+
+how to create a thread
```
Comment (sheenarbw) [inline content/topics/java-specific/concurrency/_index.md]: Is this meant to be a heading? Or is is meant to have different capitalisation?

File: content/topics/java-specific/concurrency/_index.md [hunk 10]
```diff
@@ -0,0 +1,104 @@
...
+    }
+}
+
+```
+
+#### What is Concurrency
+
+Concurrency actually means that multiple tasks can be executed in an overlapping time period. One of the tasks can begin before the preceding one is completed; however, they won’t be running at the same time. The CPU will adjust time slices per task and appropriately switch contexts. That’s why this concept is quite complicated to implement and especially debug.
+
+thre are many components/utilities in the Concurrency package, one of the commonly used one is Executor
```
Comment (sheenarbw) [inline content/topics/java-specific/concurrency/_index.md]: Spelling and capitalisation

File: content/topics/java-specific/concurrency/_index.md [hunk 11]
```diff
@@ -0,0 +1,104 @@
...
+Parallelism is the ability to execute independent tasks of a program in the same instant of time. Contrary to concurrent tasks, these tasks can run simultaneously on another processor core, another processor, or an entirely different computer that can be a distributed system. As the demand for computing speed from real-world applications increases, parallelism becomes more common and affordable.
+
+
+![concurnecy-vs-parallelism](con-vs-par.webp)
+
+
+
+#### Resources
+
+https://www.baeldung.com/cs/concurrency-vs-parallelism
```
Comment (sheenarbw) [inline content/topics/java-specific/concurrency/_index.md]: Should this be a list? Like:

- resource 1
- resource 2
- resource 3

**Review by sheenarbw — CHANGES_REQUESTED** [review]
I didn't go through absolutely all of it.

The main things are: 

- prerequisites sometimes point to things that dont exist. Please double check all the things!
- copy-writing could be better. There are a lot of places where punctuation is left out, and a few things that made me scratch my head a bit.
- some minor formatting stuff

I'm also not sure where this new content fits in the java syllabus. It would be good to update the syllabus files otherwise this content wont show up on people's boards.

I suspect that there are many people in our team who would have been able to pick up on the stuff I found. If I'm being slow to review then please just ask someone on rocketchat. You can post something on the tech-team channel or ask one of the scrummies who can assist.

File: content/projects/java-specific/functional-interface/_index.md (hunk 4 above)
Comment (ruddy-m) [reply content/projects/java-specific/functional-interface/_index.md]: not sure if this is the new hugo but this it does not print on the same line

File: content/projects/java-specific/functional-interface/_index.md [hunk 12]
```diff
@@ -0,0 +1,50 @@
...
+
+Implement the following methods, each should return a lambda expression performing a specified action:
+
+1. A method called `isOdd`: The lambda expression must return true if a number is odd or false if it is even.
+
+2. A method called `isPalindrome`: The lambda expression must return true if the number is Palindrome (12321) or false if it is not a Palindrome
+
+3. A method called `isPrime`: The lambda expression must return true if the number is Prime (3) or false if it is not a not Prime (4)
+
+Then you must get a number from the standard input which will go through all three functions and return the correct strings
```
Comment (Sbonelo01) [inline content/projects/java-specific/functional-interface/_index.md]: Maybe say:

After you have implemented the above, you must get a number from ...

File: content/projects/java-specific/functional-interface/_index.md [hunk 13]
```diff
@@ -0,0 +1,50 @@
...
+
+4. If the user enters 77 in the standard input then you should **Print!**
```
Comment (Sbonelo01) [inline content/projects/java-specific/functional-interface/_index.md]: Maybe say:

If the user enters 77 in the standard input then you should **print** the following:

File: content/projects/java-specific/functional-interface/_index.md [hunk 14]
```diff
@@ -0,0 +1,50 @@
...
+
+Then you must get a number from the standard input which will go through all three functions and return the correct strings
+
+4. If the user enters 77 in the standard input then you should **Print!** 
+
+```
+ Number is: Odd, Palindrome and NotPrime
+```
+
+Then give the user an Opportunity to enter another number
```
Comment (Sbonelo01) [inline content/projects/java-specific/functional-interface/_index.md]: Maybe drop the word "then" here

File: content/projects/java-specific/threads/_index.md [hunk 15]
```diff
@@ -0,0 +1,36 @@
...
+  soft:
+  - projects/java-specific/functional-interface
+ready: true
+submission_type: repo
+title: Java Threads
+---
+
+## The Chicken and the Farmer
+
+In this project you are going to create a `Chicken` function and a `Farmer` function each following these rules
```
Comment (Sbonelo01) [inline content/projects/java-specific/threads/_index.md]: Students might mistake the capitalization of the method names here for a requirement that they also capitalize their functions when doing the project. If this is done on purpose, please disregard my comment.

File: content/projects/java-specific/threads/_index.md [hunk 16]
```diff
@@ -0,0 +1,36 @@
...
+
+1. Have the Chicken and the Farmer run on different threads
+
+2. The Chicken should put eggs in a Basket ***(static array/list of size 1)*** and the Farmer should take eggs from the basket
+
+3. The Farmer cannot take from an empty basket
+
+4. The Chicken cannot add to a full basket
+
+5. The basket is full if it has 1 egg
```
Comment (Sbonelo01) [inline content/projects/java-specific/threads/_index.md]: Please do not forget the full stops

File: content/topics/java-specific/concurrency/_index.md [hunk 17]
```diff
@@ -0,0 +1,98 @@
+---
+content_type: topic
+ready: true
+tags:
+- threads
+- concurrency
+title: Java Multithread, Concurrency and Parallelism
+---
+
+## Have you heard?
```
Comment (Sbonelo01) [inline content/topics/java-specific/concurrency/_index.md]: I think, drop the heading and and just have the nice intro below

File: content/topics/java-specific/concurrency/_index.md [hunk 18]
```diff
@@ -0,0 +1,98 @@
...
+
+I am sure by now you have heard about java being multi-threaded, what does that mean exactly?
+
+Unlike for example JavaScript which is a single threaded language, Javaw has multiple threads to help you run your program faster.
```
Comment (Sbonelo01) [inline content/topics/java-specific/concurrency/_index.md]: some spelling and grammar mistakes.

This can be:
I am sure by now you have heard about Java being multi-threaded, what does that mean exactly?
		
Unlike for example JavaScript which is a single-threaded language, Java has multiple threads to help you run your program faster

File: content/topics/java-specific/functional-interface/_index.md [hunk 19]
```diff
@@ -0,0 +1,75 @@
...
+#### What are Lambda expressions:
+
+Introduced in Java 8 Lambda expressions are anonymous (don't have names) shortcode blocks that can take parameters and return a value just like methods.
+
+#### What are Functional Interface
+
+Functional Interface in Java is also called Single Abstract Method (SAM) interface. They can have many default and static methods but only one abstract method that is why they are also called Single Abstract Method (SAM) interface. The make it possible to create abstractions that multiple classes can use without copying and pasting code. This is especially helpful when developers need to create a complex abstraction with various methods and behaviors.
+
+
+All together Functional interface with lambda expressions help us to write smaller and cleaner code by removing a lot of boiler-plate code.
```
Comment (Sbonelo01) [inline content/topics/java-specific/functional-interface/_index.md]: Some DRY issues with the sentencing here. There are also some grammatical errors. Maybe it can be like so:

Functional Interface in Java is also called Single Abstract Method (SAM) interface. They can have many default and static methods but only one abstract method - this is why they are called SAM. They make it possible to create abstractions that multiple classes can use without copying and pasting code. This is especially helpful when developers need to create a complex abstraction with various methods and behaviors.

All together Functional interface with lambda expressions help us write smaller and cleaner code by removing a lot of boilerplate code.

File: content/topics/java-specific/functional-interface/_index.md [hunk 20]
```diff
@@ -0,0 +1,75 @@
...
+
+
+#### Functional Interface have four types
+
+![functional-interface](function-interface.png)
+
+read more about them [here](https://www.geeksforgeeks.org/functional-interfaces-java/#:~:text=Java%20SE%208%20included%20four%20main%20kinds%20of%20functional%20interfaces%20which%20can%20be%20applied%20in)
```
Comment (Sbonelo01) [inline content/topics/java-specific/functional-interface/_index.md]: Please capitalize "read"

**Review by Sbonelo01 — CHANGES_REQUESTED** [review]
The review has focused on grammar, spelling, and syllabus consistency mistakes. Please also make sure that the merge conflicts are resolved in this branch.

File: content/projects/java-specific/functional-interface/_index.md [hunk 21]
```diff
@@ -0,0 +1,51 @@
...
+4. If the user enters 77 in the standard input then you should **Print** the following: 
+
+```
+ Number is: Odd, Palindrome and NotPrime
+```
+
+The user should be able to to enter another number in the standard input.
+
+
+**Example2: User types 43 this time**
```
Comment (sheenarbw) [inline content/projects/java-specific/functional-interface/_index.md]: Is this meant to be a heading? Headings use the # symbol

File: content/projects/java-specific/functional-interface/_index.md [hunk 22]
```diff
@@ -0,0 +1,51 @@
...
+ Number is: Odd, NotPalindrome and Prime
+```
+
+***ALL FUNCTIONS MUST BE TESTED***
+
+
+#### Instructions for reviewers
+
+- Ensure that all three functions should return lambada expressions.
+
```
Comment (sheenarbw) [inline content/projects/java-specific/functional-interface/_index.md]: Why are there lank lines between the bllet points?

File: content/projects/java-specific/functional-interface/_index.md [hunk 23]
```diff
@@ -0,0 +1,51 @@
...
+**Example2: User types 43 this time**
+
+```
+ Number is: Odd, NotPalindrome and Prime
+```
+
+***ALL FUNCTIONS MUST BE TESTED***
+
+
+#### Instructions for reviewers
```
Comment (sheenarbw) [inline content/projects/java-specific/functional-interface/_index.md]: This is a level 4 heading. The last heading was level 2. 

If you have a level 2 heading then the next level is 3, not 4

File: content/projects/java-specific/functional-interface/_index.md [hunk 24]
```diff
@@ -0,0 +1,51 @@
...
+- functional-interface
+- lambda-expressions
+title: Java Lamba Expressions and Functional Interfaces
+---
+
+## What is the number
+
+Implement the following methods, each should return a lambda expression performing a specified action:
+
+1. A method called `isOdd`: The lambda expression must return true if a number is odd or false if it is even.
```
Comment (sheenarbw) [inline content/projects/java-specific/functional-interface/_index.md]: Usually, we say something about the directory structure. How should this code be structured?  

Should everything be in one file or several files? What should the class be called? 

Remember that if we need the learners top submit code in a certain way, we need to tell them about it or they wont guess correctly. 

This project looks very very automarkable so we should be hurting ourselves if we weren't specific.

File: content/projects/java-specific/threads/_index.md [hunk 25]
```diff
@@ -0,0 +1,32 @@
...
+submission_type: repo
+title: Java Threads
+---
+
+## The Chicken and the Farmer
+
+In this project you are going to create a two functions, one should be called `chicken` and the other should be called `farmer`. Both these functions should follow the following rules:
+
+1. They should run on different threads.
+
```
Comment (sheenarbw) [inline content/projects/java-specific/threads/_index.md]: should there be blank lines here?

File: content/topics/java-specific/concurrency/_index.md [hunk 26]
```diff
@@ -0,0 +1,92 @@
...
+content_type: topic
+ready: true
+tags:
+- threads
+- concurrency
+title: Java Multithread, Concurrency and Parallelism
+---
+
+
+By now, you should have heard about java being multi-threaded - but what does that mean exactly?
```
Comment (sheenarbw) [inline content/topics/java-specific/concurrency/_index.md]: should java have a capital letter? Like "Java"?

File: content/topics/java-specific/concurrency/_index.md [hunk 27]
```diff
@@ -0,0 +1,92 @@
...
+    }
+}
+
+```
+
+## What is Concurrency
+
+Concurrency actually means that multiple tasks can be executed in an overlapping time period. One of the tasks can begin before the preceding one is completed; however, they won’t be running at the same time. The CPU will adjust time slices per task and appropriately switch contexts. That’s why this concept is quite complicated to implement and especially debug.
+
+There are many components/utilities in the concurrency package, one of the commonly used one is ***Executor***
```
Comment (sheenarbw) [inline content/topics/java-specific/concurrency/_index.md]: Should we use backticks to show that Executor is a code thing? Like `Executor`?

File: content/topics/java-specific/concurrency/_index.md [hunk 28]
```diff
@@ -0,0 +1,92 @@
...
+
+**Executor:** is a set of interfaces that represents an object whose implementation executes tasks. It depends on the implementation whether the task should be run on a new thread or on a current thread. This means, we can decouple the task execution flow from the actual task execution mechanism using this interface.
```
Comment (sheenarbw) [inline content/topics/java-specific/concurrency/_index.md]: The colon on this line is funny. Either say "Executor: A set of interfaces..." OR "Executor is a set of interfaces". Only one is correct.

I'm also not sure if the sentence really makes sense overall. It might be good to double check with Ruddy...

AFAIK, `Executor` is one interface. And there are other related interfaces called `ExecutorService` and `ScheduledExecutorService`. But even though it's similar to those other interfaces it's still just one interface.

Comment (sheenarbw) [reply content/topics/java-specific/concurrency/_index.md]: This sentence is also a bit funky: "It depends on the implementation whether the task should be run on a new thread or on a current thread". Just the structure is off.

File: content/topics/java-specific/concurrency/_index.md [hunk 29]
```diff
@@ -0,0 +1,92 @@
...
+
+```
+public interface Executor {
+    void execute( Runnable command );
+}
+```
+
+In order to create an executer instance, we need to create an invoker:
```
Comment (sheenarbw) [inline content/topics/java-specific/concurrency/_index.md]: How about:

"""
Since Executor is an interface, we can't just create instances. We need to implement the Executor interface. We do this by creating an Invoker class like so:
"""

File: content/topics/java-specific/concurrency/_index.md [hunk 30]
```diff
@@ -0,0 +1,92 @@
...
+
+Now, for the execution of the task, we can use this invoker:
+
+```
+public void execute() {
+   Executor exe = new Invoker();
+   exe.execute( () -> {
+       // task to be performed
+   });
+}
```
Comment (sheenarbw) [inline content/topics/java-specific/concurrency/_index.md:71]: Overall, this whole section seems really confusing... like, where is `execute` implemented? Do you need to make the Executor interface yourself?  etc.

Is there a resource we can link to that explains this stuff well? In general it's good to use stuff that already exists as often as possible. Writing good content is hard.

File: content/topics/java-specific/functional-interface/_index.md [hunk 31]
```diff
@@ -0,0 +1,68 @@
+---
+content_type: topic
+ready: true
+tags:
+- functions
+title: Java Lambda expressions and Functional Interface
+---
+
+#### What are Lambda expressions:
```
Comment (sheenarbw) [inline content/topics/java-specific/functional-interface/_index.md]: Why are we starting off with a level 4 heading?

We have a level 1 heading ("Java Lambda expressions and Functional Interface").
The next level after 1 is 2

File: content/topics/java-specific/functional-interface/_index.md [hunk 32]
```diff
@@ -0,0 +1,68 @@
...
+
+Introduced in Java 8 Lambda expressions are anonymous(don't have names) shortcode blocks that can take parameters and return a value just like methods.
```
Comment (sheenarbw) [inline content/topics/java-specific/functional-interface/_index.md]: please revise this sentence. It needs a comma, or it needs to be 2 sentences

File: content/topics/java-specific/functional-interface/_index.md [hunk 33]
```diff
@@ -0,0 +1,68 @@
...
+
+#### What are Functional Interface
+
+Functional Interface in Java is also called Single Abstract Method (SAM) interface. They can have many default and static methods but only one abstract method - this is why they are called SAM. They make it possible to create abstractions that multiple classes can use without copying and pasting code. This is especially helpful when developers need to create a complex abstraction with various methods and behaviors.
```
Comment (sheenarbw) [inline content/topics/java-specific/functional-interface/_index.md]: Maybe "a functional interface is also called a single ...". Or "Functional interfaces are also called..." If you read the stuff aloud then you'll notice when sentence structure is strange.

File: content/topics/java-specific/functional-interface/_index.md [hunk 34]
```diff
@@ -0,0 +1,68 @@
...
+
+All together Functional interface with lambda expressions help us write smaller and cleaner code by removing a lot of boilerplate code.
```
Comment (sheenarbw) [inline content/topics/java-specific/functional-interface/_index.md]: Rather:

"Functional interfaces with lambda expressions help us to remove boilerplate code, this makes our code smaller, cleaner and faster to write"

If you read it aloud you'll get a feel for where punctuation should be. And sometimes you can flat out remove words that serve no purpose.

File: content/topics/java-specific/functional-interface/_index.md [hunk 35]
```diff
@@ -0,0 +1,68 @@
...
+
+
+#### Functional Interface have four types
+
+![functional-interface](function-interface.png)
+
+Read more about them [here](https://www.geeksforgeeks.org/functional-interfaces-java/#:~:text=Java%20SE%208%20included%20four%20main%20kinds%20of%20functional%20interfaces%20which%20can%20be%20applied%20in)
+
+- Examples of built in functional interface like **Runnable()**
```
Comment (sheenarbw) [inline content/topics/java-specific/functional-interface/_index.md]: Should this be an unordered list? Or just regular text? Or a heading?

**Review by sheenarbw — CHANGES_REQUESTED** [review]
This still needs some polish. It's definitely looking better though

File: content/projects/java-specific/functional-interface/_index.md (hunk 21 above)
Comment (Sbonelo01) [reply content/projects/java-specific/functional-interface/_index.md]: No, it's a continuation of point 4. I've fixed the sentence. It was badly structured.

File: content/projects/java-specific/functional-interface/_index.md (hunk 24 above)
Comment (Sbonelo01) [reply content/projects/java-specific/functional-interface/_index.md]: I've included a directory structure and class name. I am not 100% sure that the class name I instructed the learner to use is what Ruddy intended when creating this project.

File: content/projects/java-specific/functional-interface/_index.md [hunk 36]
```diff
@@ -0,0 +1,65 @@
...
+
+Create a class called `FunctionalInterfaceDemo` which implements methods called `isOdd`, `isPalidrome` and `isPrime`. Each of the methods should return a lambda expression.
+
+1. `isOdd`: The lambda expression must return true if a number is odd or false if it is even.
+
+2. `isPalindrome`: The lambda expression must return true if the number is Palindrome (12321) or false if it is not a Palindrome
+
+3. `isPrime`: The lambda expression must return true if the number is Prime (3) or false if it is not a not Prime (4)
+
+Once the obove has been implemented, you must get a number from the standard input which will go through all three functions and return the correct strings
```
Comment (sheenarbw) [inline content/projects/java-specific/functional-interface/_index.md]: Maybe reword this:

```
Create a main method that gets a number from the standard input and then prints a message describing the value. 

Once this message has been printed out then the user should be able to enter another number and then prints another descriptive message.

This should repeat until the user enters `q` to quit the program.

Here are some examples of descriptive messages for different numbers:

- if the user enters 77 then we print `Number is: Odd, Palindrome and NotPrime`
- If the user enters 43 then we print `Number is: Odd, NotPalindrome and Prime`
- if the user enters 44 then we print `Number is: Even, Palendrome and NotPrime`

*Hint:* Think about how you could make use of the above examples in your unit tests. 
```

File: content/projects/java-specific/introduction-to-spring-boot/consume-api/_index.md [hunk 37]
```diff
@@ -57,7 +57,7 @@ Project Setup and configuration:
 
 **Why do you need to change your jdk version for this project**
 
-1. Peace of mine, dependency hell is a thing of nightmares
+1. Peace of mind, dependency hell is a thing of nightmares
```
Comment (sheenarbw) [inline content/projects/java-specific/introduction-to-spring-boot/consume-api/_index.md:56]: question for Ruddy: If the learners are running jdk > 9 then should they downgrade?

File: content/projects/java-specific/threads/_index.md [hunk 38]
```diff
@@ -0,0 +1,28 @@
...
+submission_type: repo
+title: Java Threads
+---
+
+## The Chicken and the Farmer
+
+In this project you are going to create two functions, one should be called `chicken` and the other should be called `farmer`. Both these functions should follow the following rules:
+
+1. They should run on different threads.
+2. The chicken function should put eggs in a Basket ***(static array/list of size 1)*** and the farmer function should take eggs from the basket.
```
Comment (sheenarbw) [inline content/projects/java-specific/threads/_index.md]: inconsistent capitalization of Basket. Since it's an array/list it should be a small letter I think?

File: content/projects/java-specific/threads/_index.md [hunk 39]
```diff
@@ -0,0 +1,28 @@
...
+  - topics/java-specific/concurrency
+  - topics/java-specific/functional-interface
+  soft:
+  - projects/java-specific/functional-interface
+ready: true
+submission_type: repo
+title: Java Threads
+---
+
+## The Chicken and the Farmer
```
Comment (sheenarbw) [inline content/projects/java-specific/threads/_index.md:36]: This doesn't say anything about the directory structure. Should it? 
If the goal here is to keep it open-ended that's ok, but we might at least want to specify that they use gradle

File: content/topics/java-specific/concurrency/_index.md [hunk 40]
```diff
@@ -0,0 +1,92 @@
...
+tags:
+- threads
+- concurrency
+title: Java Multithread, Concurrency and Parallelism
+---
+
+By now, you should have heard about Java being multi-threaded - but what does that mean exactly?
+
+Unlike JavaScript for example - which is a single threaded language; Java has multiple threads to help you run your program faster. 
+
```
Comment (sheenarbw) [inline content/topics/java-specific/concurrency/_index.md:13]: If I didn't know what multi-threading was then I would be frustrated and confused for a while longer than needed. 

One trick to editing well is: Removing things and re-ordering things.

```
You may have heard about Java being multi-threaded - but what does that mean exactly?

A thread is like a separate CPU executing your application. Thus, a multithreaded application is like an application that has multiple CPUs executing different parts of the code at the same time. Multi-threaded applications have multiple threads so they can be much faster than single-threaded applications.

A thread goes through a life-cycle which can be seen in the diagram below:

...
```

File: content/topics/java-specific/concurrency/_index.md [hunk 41]
```diff
@@ -0,0 +1,92 @@
...
+## What is Multithreading
+
+Multithreading means that you have multiple threads of execution inside the same application. A thread is like a separate CPU executing your application. Thus, a multithreaded application is like an application that has multiple CPUs executing different parts of the code at the same time. 
+
+***A thread goes through a life-cycle which can be seen in the diagram below***
+
+![thread-life-cycle](Thread_Life_Cycle.jpg)
+
+### how to create a thread
+
```
Comment (sheenarbw) [inline content/topics/java-specific/concurrency/_index.md:21]: Some extra explanation is needed here. Eg:

Java can handle multiple threads, but you need to do a bit of work to set it up. This is because you, as the programmer, need to decide what each thread should be responsible for.

File: content/topics/java-specific/concurrency/_index.md [hunk 42]
```diff
@@ -0,0 +1,92 @@
...
+            System.out.println("New thread created");
+        }).start();
+    }
+}
+
+```
+
+## What is Concurrency
+
+Concurrency actually means that multiple tasks can be executed in an overlapping time period. One of the tasks can begin before the preceding one is completed; however, they won’t be running at the same time. The CPU will adjust time slices per task and appropriately switch contexts. That’s why this concept is quite complicated to implement and especially debug.
```
Comment (sheenarbw) [inline content/topics/java-specific/concurrency/_index.md]: This paragraph can do with a bit of love. Nothing major

File: content/topics/java-specific/functional-interface/_index.md [hunk 43]
```diff
@@ -0,0 +1,70 @@
...
+title: Java Lambda expressions and Functional Interface
+---
+
+## Lambda expressions
+
+Lambda expressions, introduced in Java 8, are concise and anonymous blocks of code that can take parameters and return values, similar to methods.
+
+## Functional Interface
+
+Functional Interfaces are called Single Abstract Method (SAM) interface. They can have many default and static methods but only one abstract method - this is why they are called SAM. They make it possible to create abstractions that multiple classes can use without copying and pasting code. This is especially helpful when developers need to create a complex abstraction with various methods and behaviors.
```
Comment (sheenarbw) [inline content/topics/java-specific/functional-interface/_index.md]: Functional Interfaces are called Single Abstract Method (SAM) interfaces. That last interfaces should be pleural.

File: content/topics/java-specific/functional-interface/_index.md [hunk 44]
```diff
@@ -0,0 +1,70 @@
...
+
+By combining functional interfaces with lambda expressions, we can achieve smaller and cleaner code, eliminating the need for excessive boilerplate code.
+
+#### Functional Interface have four types
```
Comment (sheenarbw) [inline content/topics/java-specific/functional-interface/_index.md]: level 4 heading after level 2 heading. This should be a level 3 heading instead.

Also, this heading seems to be in cave-man speak and the capitalisation is a bit inconsistent.

File: content/topics/java-specific/functional-interface/_index.md [hunk 45]
```diff
@@ -0,0 +1,70 @@
...
+
+![functional-interface](function-interface.png)
+
+Read more about them [here](https://www.geeksforgeeks.org/functional-interfaces-java/#:~:text=Java%20SE%208%20included%20four%20main%20kinds%20of%20functional%20interfaces%20which%20can%20be%20applied%20in)
+
+#### Example of built in functional interface:
+
+- `Runnable()`
```
Comment (sheenarbw) [inline content/topics/java-specific/functional-interface/_index.md]: what is this list item all about? It's hard to know what is intended here.

File: content/topics/java-specific/functional-interface/_index.md [hunk 46]
```diff
@@ -0,0 +1,70 @@
...
+    @Override public void run() {
+        System.out.println("New thread created");
+    }
+}).start();
+```
+
+We can use lambda expressions to re-write the code above, like so:
+
+```
+// lambda expression to create the object
```
Comment (sheenarbw) [inline content/topics/java-specific/functional-interface/_index.md]: this could do with more explanation. Does this mean that the lambda function creates a Runnable object? It looks like it does.

File: content/topics/java-specific/functional-interface/_index.md [hunk 47]
```diff
@@ -0,0 +1,70 @@
...
+
+By combining functional interfaces with lambda expressions, we can achieve smaller and cleaner code, eliminating the need for excessive boilerplate code.
+
+#### Functional Interface have four types
+
+![functional-interface](function-interface.png)
+
+Read more about them [here](https://www.geeksforgeeks.org/functional-interfaces-java/#:~:text=Java%20SE%208%20included%20four%20main%20kinds%20of%20functional%20interfaces%20which%20can%20be%20applied%20in)
+
+#### Example of built in functional interface:
```
Comment (sheenarbw) [inline content/topics/java-specific/functional-interface/_index.md]: heading level

File: content/topics/java-specific/functional-interface/_index.md [hunk 48]
```diff
@@ -0,0 +1,70 @@
...
+We can use lambda expressions to re-write the code above, like so:
+
+```
+// lambda expression to create the object
+new Thread(() -> {
+    System.out.println("New thread created");
+}).start();
+```
+
+#### Custom functional interface
```
Comment (sheenarbw) [inline content/topics/java-specific/functional-interface/_index.md]: heading level

**Review by sheenarbw — CHANGES_REQUESTED** [review]
Needs a bit of copy-editing love still. 

And there are a few things that don't make sense. 

Ruddy is around on Fridays so you can ask her for clarity on this stuff.

File: content/projects/java-specific/introduction-to-spring-boot/consume-api/_index.md (hunk 37 above)
Comment (Sbonelo01) [reply content/projects/java-specific/introduction-to-spring-boot/consume-api/_index.md:56]: @ruddy-m Please see above

File: content/projects/java-specific/threads/_index.md (hunk 39 above)
Comment (Sbonelo01) [reply content/projects/java-specific/threads/_index.md:36]: Need to get some clarity from Ruddy here

File: content/projects/java-specific/threads/_index.md (hunk 38 above)
Comment (Sbonelo01) [reply content/projects/java-specific/threads/_index.md]: it is a small letter

File: content/projects/java-specific/threads/_index.md (hunk 39 above)
Comment (ruddy-m) [reply content/projects/java-specific/threads/_index.md:36]: the directory structure is nothing special so maybe we go with
```
├── build.gradle
├── gradle
│   └── wrapper
│       ├── gradle-wrapper.jar
│       └── gradle-wrapper.properties
├── gradlew
├── gradlew.bat
├── settings.gradle
└── src
    └── main
        └── java
             └── ChickenFarm.java <------ names are important
```

File: content/projects/java-specific/introduction-to-spring-boot/consume-api/_index.md (hunk 37 above)
Comment (sheenarbw) [reply content/projects/java-specific/introduction-to-spring-boot/consume-api/_index.md:56]: I'll ask @ruddy-m in our meeting

**Review by sheenarbw — APPROVED** [review]
